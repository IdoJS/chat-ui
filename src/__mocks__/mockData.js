const MOCK_USER_NAME = 'mock_user_name';
const MOCK_AVATAR_CLASS = 'mock_avatar_class';
const MOCK_TEXT = 'mock_text';
const MOCK_SOCKET_REQ = {text: MOCK_TEXT, avatar: MOCK_AVATAR_CLASS, userName: MOCK_USER_NAME, userId: '0'};
const MOCK_SELF_MESSAGE = {text: MOCK_TEXT, avatar: MOCK_AVATAR_CLASS, userName: MOCK_USER_NAME, timestamp: 0, userId: '0'};
const MOCK_OTHER_MESSAGE = {text: MOCK_TEXT, avatar: MOCK_AVATAR_CLASS, userName: 'other_user', timestamp: 0, userId: '1'};

const MOCK_CHAT_DATA_INIT = {
  loading: false,
  timestamp: Date.now(),
  chatArr: [],
  typingArr: [],
  timeout: false
};
export {
  MOCK_USER_NAME,
  MOCK_AVATAR_CLASS,
  MOCK_TEXT,
  MOCK_SOCKET_REQ,
  MOCK_CHAT_DATA_INIT,
  MOCK_SELF_MESSAGE,
  MOCK_OTHER_MESSAGE
}