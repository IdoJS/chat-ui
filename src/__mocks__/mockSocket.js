// modify the code https://medium.freecodecamp.org/testing-socket-io-client-app-using-jest-and-react-testing-library-9cae93c070a3

let EVENTS = {};
function emit(event, ...args) {
  EVENTS[event].forEach(func => func(...args));
}
const socket = () => {
  return {
    on(event, func) {
      if (EVENTS[event]) {
        return EVENTS[event].push(func);
      }
      EVENTS[event] = [func];
    },
    emit
  }
};
export const mockIo = {
  connect() {
    return socket;
  }
};

export default mockIo;