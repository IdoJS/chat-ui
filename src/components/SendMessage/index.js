import React from 'react';
import PropTypes from 'prop-types';
import debounce from 'lodash/debounce';

const ENTER_KEY = 'Enter';


class SendMessage extends React.PureComponent {

  state = {
    inputMessage: ''
  };

  constructor(props) {
    super(props);

    this.onTextChange = this.onTextChange.bind(this);
    this.sendButton = this.sendButton.bind(this);
    this.sendEnter = this.sendEnter.bind(this);
    this.debounceTyping = debounce(this.onMessageTyping.bind(this), 500);
  }

  onTextChange(ev) {
    this.setState({
      [ev.target.name]: ev.target.value
    });

    this.debounceTyping();
  }

  onMessageTyping() {
    this.props.onMessageTyping && this.props.onMessageTyping(this.state.inputMessage.length !== 0);
  }

  sendEnter(ev) {
    if (ev.key === ENTER_KEY) {
      this.onSend()
    }
  }

  sendButton() {
    this.onSend()
  }

  onSend() {
    this.props.onMessageSend && this.props.onMessageSend({text: this.state.inputMessage});

    this.setState({
      inputMessage: ''
    });
  }

  render() {
    return (
      <div className='send_box' key={0}>
        <input className='primary_input' disabled={this.props.isLoading} name='inputMessage' type='input'
               onKeyPress={this.sendEnter}
               onChange={this.onTextChange} placeholder={this.props.isLoading ? 'Please wait...' : 'Type message...'}
               value={this.state.inputMessage}/>
        <button className='secondary_button' disabled={this.state.inputMessage.length === 0} onClick={this.sendButton}>
          send
        </button>
      </div>
    )
  }
}

SendMessage.propTypes = {
  isLoading: PropTypes.bool.isRequired,
  onMessageSend: PropTypes.func.isRequired,
  onMessageTyping: PropTypes.func.isRequired
};

SendMessage.defaultProps = {
  isLoading: false,
  userName: '',
  onMessageSend: () => {
    console.warn('Please pass onMessageSend as prop')
  },
  onMessageTyping: () => {
    console.warn('Please pass onMessageTyping as prop')
  },
};

export default SendMessage;