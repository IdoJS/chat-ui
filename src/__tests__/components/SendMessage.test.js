import React from 'react';
import {shallow} from 'enzyme';
import renderer from 'react-test-renderer';

import SendMessage from '../../../src/components/SendMessage';
import {MOCK_USER_NAME} from '../../__mocks__/mockData';

describe('SendMessage component', () => {

  it('Should render the SendMessage view && isLoading is false', () => {
    const wrapper = shallow(<SendMessage isLoading={false} userName={MOCK_USER_NAME}
                                         onMessageSend={() => {
                                         }}
                                         onMessageTyping={() => {
                                         }}/>);
    const btn = wrapper.find('button');
    const input = wrapper.find('input');

    const msg = 'test';

    expect(btn).toHaveLength(1);
    expect(input).toHaveLength(1);

    expect(btn.props()['disabled']).toBeTruthy() // no text in input
    expect(input.props()['disabled']).toBeFalsy(); // isloading === false
    expect(input.props()['placeholder']).toMatch('Type message...'); // isloading === false

    input.simulate('change', {
      target: {
        value: msg,
        name: 'inputMessage'
      }
    });

    expect(wrapper.instance().state.inputMessage).toEqual(msg);

  });

  it('Should render the SendMessage view && isLoading is true', () => {
    const wrapper = shallow(<SendMessage isLoading={true} userName={MOCK_USER_NAME}
                                         onMessageSend={() => {
                                         }}
                                         onMessageTyping={() => {
                                         }}/>);
    const btn = wrapper.find('button');
    const input = wrapper.find('input');

    expect(btn.props()['disabled']).toBeTruthy(); // no text in input
    expect(input.props()['disabled']).toBeTruthy(); // isloading === true
    expect(input.props()['placeholder']).toMatch('Please wait...'); // isloading === true


  });

  it('Simulate send via button', (done) => {
    const msg = 'test';

    const wrapper = shallow(<SendMessage isLoading={false} userName={MOCK_USER_NAME}
                                         onMessageSend={(data) => {
                                           expect(data).toMatchObject({'text': msg});
                                           done();
                                         }}
                                         onMessageTyping={() => {
                                         }}/>);
    const btn = wrapper.find('button');
    const input = wrapper.find('input');


    input.simulate('change', {
      target: {
        value: msg,
        name: 'inputMessage'
      }
    });


    btn.simulate('click',{
      preventDefault: () => {
      }
    });

  });

  it('Simulate send via Enter Key', (done) => {
    const msg = 'test';

    const wrapper = shallow(<SendMessage isLoading={false} userName={MOCK_USER_NAME}
                                         onMessageSend={(data) => {
                                           expect(data).toMatchObject({'text': msg});
                                           done();
                                         }}
                                         onMessageTyping={() => {
                                         }}/>);
    const input = wrapper.find('input');

    input.simulate('change', {
      target: {
        value: msg,
        name: 'inputMessage'
      }
    });

    input.simulate('keyPress', {
      key : 'Enter'
    })

  });

  it('SnapShot rendering SendMessage', () => {
    const wrapper = shallow(<SendMessage/>);
    const tree = renderer.create(wrapper).toJSON();
    expect(tree).toMatchSnapshot();
  });

});