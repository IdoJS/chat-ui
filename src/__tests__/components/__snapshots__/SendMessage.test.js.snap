// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SendMessage component SnapShot rendering SendMessage 1`] = `
<div
  className="send_box"
>
  <input
    className="primary_input"
    disabled={false}
    name="inputMessage"
    onChange={[Function]}
    onKeyPress={[Function]}
    placeholder="Type message..."
    type="input"
    value=""
  />
  <button
    className="secondary_button"
    disabled={true}
    onClick={[Function]}
  >
    send
  </button>
</div>
`;
